File: pkg/engine/engine_test.go
package engine

import (
	"bytes"
	"errors"
	"strings"
	"testing"
	"time"

	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game/elements"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/stack"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles/binarytiles"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles/tiletemplates"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tilesets"
)

type testResponse struct {
	BaseResponse
}
type testRequest struct {
	GameID        int
	RequiresWrite bool
	executeFunc   func(req *testRequest, game *game.Game) Response
}

func (req *testRequest) gameID() int {
	return req.GameID
}

func (req *testRequest) requiresWrite() bool {
	return req.RequiresWrite
}

func (req *testRequest) execute(game *game.Game) Response {
	if req.executeFunc != nil {
		return req.executeFunc(req, game)
	}
	return &testResponse{BaseResponse{gameID: req.gameID()}}
}

type testResponseWithRemovableGame struct {
	BaseResponse
	canRemoveGameFunc func(resp *testResponseWithRemovableGame) bool
}

func (resp *testResponseWithRemovableGame) canRemoveGame() bool {
	if resp.canRemoveGameFunc != nil {
		return resp.canRemoveGameFunc(resp)
	}
	return false
}

func TestFullGame(t *testing.T) {
	engine, err := StartGameEngine(4, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}
	tileSet := tilesets.StandardTileSet()

	gameWithID, err := engine.GenerateOrderedGame(tileSet)
	if err != nil {
		t.Fatal(err.Error())
	}
	game, gameID := gameWithID.Game, gameWithID.ID

	t.Logf("before loop: %s\n", time.Now())
	for i, expectedTile := range tileSet.Tiles {
		if !game.CurrentTile.Equals(expectedTile) {
			t.Fatalf(
				"expected %v-th tile to be %#v, got %#v instead",
				i,
				expectedTile,
				game.CurrentTile,
			)
		}

		t.Logf(
			"iteration %v start: %v\n", i, binarytiles.FromTile(game.CurrentTile),
		)
		legalMovesReq := &GetLegalMovesRequest{
			BaseGameID: gameID, TileToPlace: game.CurrentTile,
		}
		legalMovesResp := engine.SendGetLegalMovesBatch(
			[]*GetLegalMovesRequest{legalMovesReq},
		)[0]
		if legalMovesResp.Err() != nil {
			t.Fatal(legalMovesResp.Err().Error())
		}
		t.Logf("iteration %v got moves\n", i)

		move := legalMovesResp.Moves[0].Move
		t.Logf(
			"iteration %v selecting move: %v at position %v\n",
			i,
			binarytiles.FromPlacedTile(move),
			move.Position,
		)
		playTurnReq := &PlayTurnRequest{GameID: gameID, Move: move}
		playTurnResp := engine.SendPlayTurnBatch([]*PlayTurnRequest{playTurnReq})[0]
		if playTurnResp.Err() != nil {
			t.Fatal(playTurnResp.Err().Error())
		}
		t.Logf("iteration %v played turn\n", i)

		game = playTurnResp.Game
		gameID = playTurnResp.GameID()
		t.Logf("iteration %v end: %s\n", i, time.Now())
	}

	if len(game.CurrentTile.Features) != 0 {
		t.Fatalf("expected current tile to be nil, got %#v instead", game.CurrentTile)
	}
}

func TestConcurrentReadRequests(t *testing.T) {
	engine, err := StartGameEngine(4, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}
	tileSet := tilesets.StandardTileSet()

	gameWithID, err := engine.GenerateGame(tileSet)
	if err != nil {
		t.Fatal(err.Error())
	}
	game, gameID := gameWithID.Game, gameWithID.ID

	legalMovesReq := &GetLegalMovesRequest{
		BaseGameID: gameID, TileToPlace: game.CurrentTile,
	}
	legalMovesResp := engine.SendGetLegalMovesBatch(
		[]*GetLegalMovesRequest{legalMovesReq},
	)[0]
	if legalMovesResp.Err() != nil {
		t.Fatal(legalMovesResp.Err().Error())
	}

	requests := make([]*GetRemainingTilesRequest, len(legalMovesResp.Moves))
	for i, moveWithState := range legalMovesResp.Moves {
		requests[i] = &GetRemainingTilesRequest{
			BaseGameID: gameID, StateToCheck: moveWithState.State,
		}
	}
	responses := engine.SendGetRemainingTilesBatch(requests)

	for _, resp := range responses {
		err := resp.Err()
		if err != nil {
			t.Fatal(err)
		}
	}
}

func TestGameEngineDoubleCloseDoesNotPanic(t *testing.T) {
	engine, err := StartGameEngine(4, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}
	engine.Close()
	engine.Close()
}

func TestGameEngineCloneGameReturnsIndependentGames(t *testing.T) {
	engine, err := StartGameEngine(1, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	buf := bytes.Buffer{}
	engine.appLogger.SetOutput(&buf)

	g, err := engine.GenerateGame(tilesets.StandardTileSet())
	if err != nil {
		t.Fatal(err.Error())
	}

	_, err = engine.CloneGame(g.ID, 3)
	if err != nil {
		t.Fatal(err.Error())
	}

	logs := buf.String()
	if len(logs) > 0 {
		t.Fatalf(
			"expected logs to be empty but they weren't. full logs below:\n%v",
			logs,
		)
	}

	engine.Close()
}

func TestGameEngineSendPlayTurnBatchDoesNotWarnAboutRemovedChildren(t *testing.T) {
	engine, err := StartGameEngine(1, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	buf := bytes.Buffer{}
	engine.appLogger.SetOutput(&buf)

	g, err := engine.GenerateGame(tilesets.StandardTileSet())
	if err != nil {
		t.Fatal(err.Error())
	}

	ids, err := engine.SubCloneGame(g.ID, 15)
	if err != nil {
		t.Fatal(err.Error())
	}

	engine.DeleteGames(ids)

	req := &PlayTurnRequest{GameID: g.ID, Move: g.Game.ValidTilePlacements[0]}
	engine.SendPlayTurnBatch([]*PlayTurnRequest{req})

	logs := buf.String()
	if len(logs) > 0 {
		t.Fatalf(
			"expected logs to be empty but they weren't. full logs below:\n%v",
			logs,
		)
	}

	engine.Close()
}

func TestGameEngineSendPlayTurnBatchRemovesFinishedGames(t *testing.T) {
	engine, err := StartGameEngine(1, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	buf := bytes.Buffer{}
	engine.appLogger.SetOutput(&buf)

	g, err := engine.GenerateGame(
		tilesets.TileSet{
			StartingTile: tiletemplates.SingleCityEdgeStraightRoads(),
			Tiles:        []tiles.Tile{},
		},
	)
	if err != nil {
		t.Fatal(err.Error())
	}

	ids, err := engine.SubCloneGame(g.ID, 3)
	if err != nil {
		t.Fatal(err.Error())
	}

	for _, gameID := range []int{ids[0], g.ID} {
		if _, exists := engine.games[gameID]; !exists {
			t.Fatal("expected game to exist before final round")
		}

		req := &PlayTurnRequest{GameID: gameID}
		resp := engine.SendPlayTurnBatch([]*PlayTurnRequest{req})[0]
		if !errors.Is(resp.err, stack.ErrStackOutOfBounds) {
			t.Fatal(err.Error())
		}

		if _, exists := engine.games[gameID]; exists {
			t.Fatal("expected game to not exist after final round")
		}
	}

	engine.Close()
}

func TestGameEngineSendBatchReceivesCorrectResponsesAfterWorkerRequests(t *testing.T) {
	engine, err := StartGameEngine(4, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	requestCount := 100
	requests := make([]Request, 0, requestCount)
	for range requestCount {
		g, err := engine.GenerateGame(tilesets.StandardTileSet())
		if err != nil {
			t.Fatal(err.Error())
		}

		req := &testRequest{GameID: g.ID}
		requests = append(requests, req)
	}

	responses := engine.sendBatch(requests)
	for i, resp := range responses {
		err := resp.Err()
		if err != nil {
			t.Fatal(err.Error())
		}
		expected := requests[i].gameID()
		actual := resp.GameID()
		if actual != expected {
			t.Fatalf("expected %v game ID, got %v instead", expected, actual)
		}
	}
	engine.Close()
}

func TestGameEngineSendBatchReturnsFailureWhenGameIDNotFound(t *testing.T) {
	engine, err := StartGameEngine(4, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	requests := make([]Request, 0, 2)
	g, err := engine.GenerateGame(tilesets.StandardTileSet())
	if err != nil {
		t.Fatal(err.Error())
	}

	successfulReq := &testRequest{GameID: g.ID}
	requests = append(requests, successfulReq)

	wrongID := g.ID + 2
	failingReq := &testRequest{GameID: wrongID}
	requests = append(requests, failingReq)
	responses := engine.sendBatch(requests)

	// successful req
	err = responses[0].Err()
	if err != nil {
		t.Fatal(err.Error())
	}
	expected := g.ID
	actual := responses[0].GameID()
	if expected != actual {
		t.Fatalf("expected %v game ID, got %v instead", expected, actual)
	}

	// failing req
	err = responses[1].Err()
	if err == nil {
		t.Fatal("expected error to occur")
	}
	if !errors.Is(err, ErrGameNotFound) {
		t.Fatal(err.Error())
	}
	expected = wrongID
	actual = responses[1].GameID()
	if expected != actual {
		t.Fatalf("expected %v game ID, got %v instead", expected, actual)
	}
	engine.Close()
}

func TestGameEngineSendBatchReturnsFailuresWhenCommunicatorClosed(t *testing.T) {
	engine, err := StartGameEngine(4, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	requestCount := 5
	requests := make([]Request, 0, requestCount)
	for range requestCount {
		g, err := engine.GenerateGame(tilesets.StandardTileSet())
		if err != nil {
			t.Fatal(err.Error())
		}

		req := &testRequest{GameID: g.ID}
		requests = append(requests, req)
	}
	engine.Close()

	responses := engine.sendBatch(requests)
	for i, resp := range responses {
		err := resp.Err()
		if err == nil {
			t.Fatal("expected error to occur")
		}
		if !errors.Is(err, ErrCommunicatorClosed) {
			t.Fatal(err.Error())
		}
		expected := requests[i].gameID()
		actual := resp.GameID()
		if actual != expected {
			t.Fatalf("expected %v game ID, got %v instead", expected, actual)
		}
	}
}

func TestGameEngineSendBatchReturnsExecutionPanicErrorOnPanicInWorker(t *testing.T) {
	engine, err := StartGameEngine(4, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	requestCount := 5
	requests := make([]Request, 0, requestCount)
	for range requestCount {
		g, err := engine.GenerateGame(tilesets.StandardTileSet())
		if err != nil {
			t.Fatal(err.Error())
		}

		req := &testRequest{GameID: g.ID}
		requests = append(requests, req)
	}
	req := requests[0].(*testRequest)
	req.executeFunc = func(*testRequest, *game.Game) Response {
		panic("panic during function execution")
	}

	responses := engine.sendBatch(requests)
	for i, resp := range responses {
		err := resp.Err()
		if i == 0 {
			if err == nil {
				t.Fatal("expected error to occur")
			}
			_, panicOccured := err.(*ExecutionPanicError)
			if !panicOccured {
				t.Fatal(err)
			}
		} else if err != nil {
			t.Fatal(err)
		}
		expected := requests[i].gameID()
		actual := resp.GameID()
		if actual != expected {
			t.Fatalf("expected %v game ID, got %v instead", expected, actual)
		}
	}
	engine.Close()
}

func TestGameEngineSendBatchReturnsExecutionPanicErrorOnPanicInSendBatchChildGoroutine(t *testing.T) {
	engine, err := StartGameEngine(4, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	requestCount := 5
	requests := make([]Request, 0, requestCount)
	for range requestCount {
		g, err := engine.GenerateGame(tilesets.StandardTileSet())
		if err != nil {
			t.Fatal(err.Error())
		}

		req := &testRequest{GameID: g.ID}
		requests = append(requests, req)
	}
	req := requests[0].(*testRequest)
	req.executeFunc = func(req *testRequest, _ *game.Game) Response {
		return &testResponseWithRemovableGame{
			BaseResponse: BaseResponse{gameID: req.gameID()},
			canRemoveGameFunc: func(*testResponseWithRemovableGame) bool {
				panic("panic during function execution")
			},
		}
	}

	responses := engine.sendBatch(requests)
	for i, resp := range responses {
		err := resp.Err()
		if err == nil {
			t.Fatal("expected error to occur")
		}
		_, panicOccured := err.(*ExecutionPanicError)
		if !panicOccured {
			t.Fatal(err)
		}
		expected := requests[i].gameID()
		actual := resp.GameID()
		if actual != expected {
			t.Fatalf("expected %v game ID, got %v instead", expected, actual)
		}
	}
	engine.Close()
}

type testRequestPanickingOnRequiresWriteCall struct {
	GameID int
}

func (req *testRequestPanickingOnRequiresWriteCall) gameID() int {
	return req.GameID
}

func (req *testRequestPanickingOnRequiresWriteCall) requiresWrite() bool {
	panic("panic during requiresWrite()")
}

func (req *testRequestPanickingOnRequiresWriteCall) execute(_ *game.Game) Response {
	panic("this should not ever be reached")
}

func TestGameEngineSendBatchReturnsExecutionPanicErrorOnPanicInMainThread(t *testing.T) {
	engine, err := StartGameEngine(4, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	requestCount := 5
	requests := make([]Request, 0, requestCount)
	for range requestCount {
		g, err := engine.GenerateGame(tilesets.StandardTileSet())
		if err != nil {
			t.Fatal(err.Error())
		}

		req := &testRequest{GameID: g.ID}
		requests = append(requests, req)
	}
	requests[0] = &testRequestPanickingOnRequiresWriteCall{requests[1].gameID()}

	responses := engine.sendBatch(requests)
	for i, resp := range responses {
		err := resp.Err()
		if err == nil {
			t.Fatal("expected error to occur")
		}
		_, panicOccured := err.(*ExecutionPanicError)
		if !panicOccured {
			t.Fatal(err)
		}
		expected := requests[i].gameID()
		actual := resp.GameID()
		if actual != expected {
			t.Fatalf("expected %v game ID, got %v instead", expected, actual)
		}
	}
	engine.Close()
}

func TestGameEngineSendBatchReturnsExecutionPanicErrorOnPanicEverywhere(t *testing.T) {
	engine, err := StartGameEngine(3, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}

	requestCount := 5
	requests := make([]Request, 0, requestCount)
	for range requestCount {
		g, err := engine.GenerateGame(tilesets.StandardTileSet())
		if err != nil {
			t.Fatal(err.Error())
		}

		req := &testRequest{GameID: g.ID}
		requests = append(requests, req)
	}
	req := requests[0].(*testRequest)
	req.executeFunc = func(req *testRequest, _ *game.Game) Response {
		return &testResponseWithRemovableGame{
			BaseResponse: BaseResponse{gameID: req.gameID()},
			canRemoveGameFunc: func(*testResponseWithRemovableGame) bool {
				panic("panic during function execution")
			},
		}
	}
	requests[1] = &testRequestPanickingOnRequiresWriteCall{requests[1].gameID()}
	req = requests[2].(*testRequest)
	req.executeFunc = func(*testRequest, *game.Game) Response {
		panic("panic during function execution")
	}

	responses := engine.sendBatch(requests)
	for i, resp := range responses {
		err := resp.Err()
		if err == nil {
			t.Fatal("expected error to occur")
		}
		_, panicOccured := err.(*ExecutionPanicError)
		if !panicOccured {
			t.Fatal(err)
		}
		msg := err.Error()
		if strings.Count(msg, "stack trace") < 2 {
			t.Fatalf(
				"expected at least 2 stack traces, got following message instead:\n%v",
				msg,
			)
		}
		expected := requests[i].gameID()
		actual := resp.GameID()
		if actual != expected {
			t.Fatalf("expected %v game ID, got %v instead", expected, actual)
		}
	}
	engine.Close()
}

func TestGenerateOrderedGame(t *testing.T) {
	engine, err := StartGameEngine(3, t.TempDir())
	if err != nil {
		t.Fatal(err.Error())
	}
	_, err = engine.GenerateOrderedGame(tilesets.MiniTileSet())
	if err != nil {
		t.Fatal(err.Error())
	}

	engine.Close()
}

func findMeeple(tile elements.PlacedTile) {
	found := false
	for _, feature := range tile.Features {
		if feature.Meeple.Type != elements.NoneMeeple {
			fmt.Printf("Found meeple on side: %#v and featuretype: %#v\n", feature.Sides, feature.FeatureType)
			found = true
		}
	}

	if !found {
		println("Meeple not found")
	}
}

File: python_bindings/tests/engine_test.py
import logging
from pathlib import Path

import pytest
from pytest import approx

from carcassonne_engine import GameEngine, models, tiletemplates
from carcassonne_engine._bindings.elements import MeepleType
from carcassonne_engine._bindings.side import Side
from carcassonne_engine._bindings.feature import Type as FeatureType
from carcassonne_engine.models import Position
from carcassonne_engine.tilesets import TileSet
from carcassonne_engine.requests import (
    GetLegalMovesRequest,
    GetMidGameScoreRequest,
    GetRemainingTilesRequest,
    PlayTurnRequest,
)
from carcassonne_engine.tilesets import TileSet, standard_tile_set
from carcassonne_engine.utils import format_binary_tile_bits
from tests.utils import make_turn, TurnParams

log = logging.getLogger(__name__)


def test_full_game(tmp_path: Path) -> None:
    engine = GameEngine(4, tmp_path)
    tile_set = standard_tile_set()

    game_id, game = engine.generate_ordered_game(tile_set)

    for i, expected_tile in enumerate(tile_set):
        assert game.current_tile == expected_tile
        log.info(
            "iteration %s start: %s",
            i,
            format_binary_tile_bits(game.current_tile.to_bits()),
        )
        legal_moves_req = GetLegalMovesRequest(
            base_game_id=game_id, tile_to_place=game.current_tile
        )
        (legal_moves_resp,) = engine.send_get_legal_moves_batch([legal_moves_req])
        assert legal_moves_resp.exception is None
        assert legal_moves_resp.moves is not None
        log.info("iteration %s got moves", i)

        move = legal_moves_resp.moves[0].move
        log.info(
            "iteration %s selecting move: %s at position %s",
            i,
            format_binary_tile_bits(move.to_bits()),
            move.position,
        )
        play_turn_req = PlayTurnRequest(game_id=game_id, move=move)
        (play_turn_resp,) = engine.send_play_turn_batch([play_turn_req])
        assert play_turn_resp.exception is None
        assert play_turn_resp.game is not None
        log.info("iteration %s played turn", i)

        game = play_turn_resp.game
        game_id = play_turn_resp.game_id
        log.info("iteration %s end", i)

    assert game.current_tile is None


def test_concurrent_read_requests(tmp_path: Path) -> None:
    engine = GameEngine(4, tmp_path)
    tile_set = standard_tile_set()

    game_id, game = engine.generate_game(tile_set)
    assert game.current_tile is not None

    legal_moves_req = GetLegalMovesRequest(
        base_game_id=game_id, tile_to_place=game.current_tile
    )
    (legal_moves_resp,) = engine.send_get_legal_moves_batch([legal_moves_req])
    assert legal_moves_resp.exception is None
    assert legal_moves_resp.moves is not None

    requests = []
    for move_with_state in legal_moves_resp.moves:
        requests.append(
            GetRemainingTilesRequest(
                base_game_id=game_id, state_to_check=move_with_state.state
            )
        )
    responses = engine.send_get_remaining_tiles_batch(requests)

    for resp in responses:
        assert resp.exception is None


def test_game_engine_send_batch_receives_correct_responses_after_worker_requests(
    tmp_path: Path,
) -> None:
    engine = GameEngine(4, tmp_path)
    request_count = 100

    tile_set = standard_tile_set()
    games = [engine.generate_game(tile_set) for _ in range(request_count)]
    requests = [
        PlayTurnRequest(
            game_id=generated_game.id,
            move=models.PlacedTile(generated_game.game._go_obj.ValidTilePlacements[0]),
        )
        for generated_game in games
    ]

    responses = engine.send_play_turn_batch(requests)
    for idx, resp in enumerate(responses):
        assert resp.exception is None
        expected = requests[idx].game_id
        actual = resp.game_id
        assert actual == expected
    engine.close()


def test_game_engine_send_batch_returns_failure_when_game_id_not_found(
    tmp_path: Path,
) -> None:
    engine = GameEngine(4, tmp_path)
    requests = []

    game_id, game = engine.generate_game(standard_tile_set())

    successful_req = PlayTurnRequest(
        game_id=game_id,
        move=models.PlacedTile(game._go_obj.ValidTilePlacements[0]),
    )
    requests.append(successful_req)

    wrong_id = game_id + 2
    failed_req = PlayTurnRequest(
        game_id=wrong_id,
        move=models.PlacedTile(game._go_obj.ValidTilePlacements[0]),
    )
    requests.append(failed_req)

    responses = engine.send_play_turn_batch(requests)
    assert responses[0].exception is None
    expected = game_id
    actual = responses[0].game_id
    assert expected == actual

    assert isinstance(responses[1].exception, Exception)
    expected = wrong_id
    actual = responses[1].game_id
    assert expected == actual

    engine.close()


def test_game_engine_send_batch_raises_when_communicator_closed(
    tmp_path: Path,
) -> None:
    engine = GameEngine(4, tmp_path)
    request_count = 5

    tile_set = standard_tile_set()
    games = [engine.generate_game(tile_set) for _ in range(request_count)]
    requests = [
        PlayTurnRequest(
            game_id=generated_game.id,
            move=models.PlacedTile(generated_game.game._go_obj.ValidTilePlacements[0]),
        )
        for generated_game in games
    ]
    engine.close()

    with pytest.raises(RuntimeError):
        engine.send_play_turn_batch(requests)


def test_game_engine_send_get_remaining_tiles_batch_returns_remaining_tiles(
    tmp_path: Path,
) -> None:
    t1 = tiletemplates.monastery_with_single_road()
    t2 = tiletemplates.roads_turn()
    tiles = [t1, t2, t1]
    tile_set = TileSet.from_tiles(
        tiles,
        starting_tile=tiletemplates.single_city_edge_straight_roads(),
    )

    with GameEngine(1, tmp_path) as engine:
        game_id, game = engine.generate_game(tile_set)

        request = GetRemainingTilesRequest(base_game_id=game_id)
        (resp,) = engine.send_get_remaining_tiles_batch([request])
        assert resp.exception is None
        assert resp.tile_probabilities is not None

    for tile_prob in resp.tile_probabilities:
        for tile in tiles:
            if tile_prob.tile == tile:
                assert tile_prob.probability == approx(tiles.count(tile) / len(tiles))
                break
        else:
            assert (
                False
            ), f"could not find a tile matching the move probability {tile_prob}"


def test_game_engine_send_get_legal_moves_batch_returns_no_duplicates(
    tmp_path: Path,
) -> None:
    tile = tiletemplates.monastery_without_roads()
    tile_set = TileSet.from_tiles(
        [tile],
        starting_tile=tiletemplates.single_city_edge_straight_roads(),
    )

    with GameEngine(1, tmp_path) as engine:
        game_id, game = engine.generate_game(tile_set)

        request = GetLegalMovesRequest(base_game_id=game_id, tile_to_place=tile)
        (resp,) = engine.send_get_legal_moves_batch([request])
        assert resp.exception is None
        assert resp.moves is not None

    # Monastery with no roads is symmetrical both horizontally and vertically
    # and there's just one valid position (below starting tile, i.e. (0, -1))
    # so there's three legal moves: without meeple and with meeple on each of
    # the 2 features
    assert len(resp.moves) == 3
    for move_with_state in resp.moves:
        assert move_with_state.move.position.x == 0
        assert move_with_state.move.position.y == -1
        assert tile.exact_equals(move_with_state.move.to_tile())


def test_game_engine_send_get_legal_moves_batch_returns_all_legal_rotations(
    tmp_path: Path,
) -> None:
    tile = tiletemplates.monastery_with_single_road()
    tile_set = TileSet.from_tiles(
        [tile],
        # non-default starting tile - limits number of possible positions to one
        starting_tile=tiletemplates.three_city_edges_connected(),
    )

    with GameEngine(1, tmp_path) as engine:
        game_id, game = engine.generate_game(tile_set)

        request = GetLegalMovesRequest(base_game_id=game_id, tile_to_place=tile)
        (resp,) = engine.send_get_legal_moves_batch([request])
        assert resp.exception is None
        assert resp.moves is not None

    # Monastery with single road can only be placed at (0, -1)
    # but in 3 different orientations (only field connected with road is invalid)
    # For each orientation, there are 4 valid meeple placements:
    # - no meeple
    # - meeple on monastery
    # - meeple on field
    # - meeple on road
    expected_move_count = 1 * 3 * 4

    assert len(resp.moves) == expected_move_count
    for move_state in resp.moves:
        assert move_state.move.position.x == 0
        assert move_state.move.position.y == -1

    # starting tile has a field at the bottom so expected values are:
    expected_tiles = [
        # road at the bottom
        tile,
        # road on the left
        models.Tile(tile._go_obj.Rotate(1)),
        # road on the right
        models.Tile(tile._go_obj.Rotate(3)),
    ]
    move_idx = 0
    for expected in expected_tiles:
        for _ in range(4):
            move_state = resp.moves[move_idx]
            assert expected.exact_equals(move_state.move.to_tile())
            move_idx += 1


def test_mid_game_score_request_at_start(tmp_path: Path) -> None:
    engine = GameEngine(4, tmp_path)
    tile_set = standard_tile_set()

    game_id, game = engine.generate_game(tile_set)

    mid_game_score_request = GetMidGameScoreRequest(
        base_game_id=game_id,
    )
    (mid_game_score_response,) = engine.send_get_mid_game_score_batch(
        [mid_game_score_request]
    )

    assert mid_game_score_response.player_scores == {1: 0, 2: 0}

def test_mid_game_score_request_mid_game(tmp_path: Path) -> None:
    engine = GameEngine(4, tmp_path)
    tiles = [
        tiletemplates.four_city_edges_connected_shield(),
        tiletemplates.straight_roads(),
        tiletemplates.straight_roads(),
    ]
    tile_set = TileSet.from_tiles(tiles=tiles,starting_tile= tiletemplates.single_city_edge_straight_roads())

    game_id, game = engine.generate_ordered_game(tile_set)

    # play turns
    turn_params = TurnParams(
        pos=Position(x=0, y=1),
        tile=tiletemplates.four_city_edges_connected_shield(),
        meepleType=MeepleType.NormalMeeple,
        side=Side.Left,
        featureType=FeatureType.City,
    )
    game_id, game = make_turn(engine, game, game_id, turn_params)

    turn_params = TurnParams(
        pos=Position(x=1, y=0),
        tile=tiletemplates.straight_roads(),
        meepleType=MeepleType.NormalMeeple,
        side=Side.Left,
        featureType=FeatureType.Road,
    )
    game_id, game = make_turn(engine, game, game_id, turn_params)

    # check scores
    mid_game_score_request = GetMidGameScoreRequest(
        base_game_id=game_id,
    )
    (mid_game_score_response,) = engine.send_get_mid_game_score_batch(
        [mid_game_score_request]
    )

    assert mid_game_score_response.player_scores == {1: 3, 2: 2}
